app.js

```diff
@@ -154,31 +154,63 @@ function addEventListener(indicator, emitter, event, f) {
 }
 
 /*
-  Check the referenced database exists and that the credentials used have
+  Check the backup database exists and that the credentials used have
   visibility. Callback with a fatal error if there is a problem with the DB.
   @param {string} db - database object
   @param {function(err)} callback - error is undefined if DB exists
 */
-function proceedIfDbValid(db, callback) {
+function proceedIfBackupDbValid(db, callback) {
   db.service.headDatabase({ db: db.db }).then(() => callback()).catch(err => {
-    err = error.convertResponseError(err, function(err) {
-      if (err && err.status === 404) {
-        // Override the error type and mesasge for the DB not found case
-        var msg = `Database ${db.url}` +
-        `${db.db} does not exist. ` +
-        'Check the URL and database name have been specified correctly.';
-        var noDBErr = new Error(msg);
-        noDBErr.name = 'DatabaseNotFound';
-        return noDBErr;
-      } else {
-        // Delegate to the default error factory if it wasn't a 404
-        return error.convertResponseError(err);
-      }
-    });
+    err = error.convertResponseError(err, err => parseIfDbValidResponseError(db, err));
+    callback(err);
+  });
+}
+
+/*
+  Check that the restore database exists, is new and is empty. Also verify that the credentials used have
+  visibility. Callback with a fatal error if there is a problem with the DB.
+  @param {string} db - database object
+  @param {function(err)} callback - error is undefined if DB exists, new and empty
+*/
+function proceedIfRestoreDbValid(db, callback) {
+  db.service.getDatabaseInformation({ db: db.db }).then(response => {
+    const { doc_count: docCount, doc_del_count: deletedDocCount } = response.result;
+    // The system databases can have a validation ddoc(s) injected in them on creation.
+    // This sets the doc count off, so we just complitely exclude the system databases from this check.
+    // The assumption here is that users restoring system databases know what they are doing.
+    if (!db.db.startsWith("_") && (docCount !== 0 || deletedDocCount !== 0)) {
+      var notEmptyDBErr = new Error(`Target database ${db.url}${db.db} is not empty.`);
+      notEmptyDBErr.name = 'DatabaseNotEmpty';
+      callback(notEmptyDBErr);
+    } else {
+      callback();
+    }
+  }).catch(err => {
+    err = error.convertResponseError(err, err => parseIfDbValidResponseError(db, err));
     callback(err);
   });
 }
 
+/*
+  Convert the database validation response error to a special DatabaseNotFound error
+  in case the database is missing. Otherwise delegate to the default error factory.
+  @param {object} db - database object
+  @param {object} err - HTTP response error
+*/
+function parseIfDbValidResponseError(db, err) {
+  if (err && err.status === 404) {
+    // Override the error type and mesasge for the DB not found case
```

Comment:
When I skim through this PR I found a typo:
```suggestion
    // Override the error type and message for the DB not found case
```

Reply:
Ah, thank you! This is a prior art, not my change, so haven't noticed that 😂